Marlin/src/HAL/STM32/HardwareSerial.cpp

```diff
@@ -0,0 +1,445 @@
+  void serialEvent2() __attribute__((weak));
+#endif
+
+#ifdef USING_HW_SERIAL3
+  HAL_HardwareSerial Serial3(USART3);
+  void serialEvent3() __attribute__((weak));
+#endif
+
+#ifdef USING_HW_SERIAL4
+  #ifdef USART4
+    HAL_HardwareSerial HSerial4(USART4);
+  #else
+    HAL_HardwareSerial HSerial4(UART4);
+  #endif
+  void serialEvent4() __attribute__((weak));
+#endif
+
+#ifdef USING_HW_SERIAL5
+  #ifdef USART5
+    HAL_HardwareSerial HSerial5(USART5);
+  #else
+    HAL_HardwareSerial HSerial5(UART5);
+  #endif
+  void serialEvent5() __attribute__((weak));
+#endif
+
+#ifdef USING_HW_SERIAL6
+  #ifdef USART6
+    HAL_HardwareSerial HSerial5(USART6);
+  #else
+    HAL_HardwareSerial HSerial5(UART6);
+  #endif
+  void serialEvent5() __attribute__((weak));
+#endif
+
+// Constructors ////////////////////////////////////////////////////////////////
+
+HAL_HardwareSerial::HAL_HardwareSerial(void *peripheral) {
+  if (peripheral == USART1) {
+    setRx(PIN_SERIAL1_RX);
+    setTx(PIN_SERIAL1_TX);
+    _uart_index = 0;
+    #if defined(STM32F2xx) || defined(STM32F4xx) || defined(STM32F7xx)
+      RX_DMA = { USART1, RCC_AHB1Periph_DMA2, 4, DMA2_Stream2 };
+    #endif
+    #if defined(STM32F0xx) || defined(STM32F1xx)
+      RX_DMA = { USART1, RCC_AHBPeriph_DMA1, DMA1, DMA1_Channel5 };
+    #endif
+  }
+  else if (peripheral == USART2) {
+    setRx(PIN_SERIAL2_RX);
+    setTx(PIN_SERIAL2_TX);
+    _uart_index = 1;
+    #if defined(STM32F2xx) || defined(STM32F4xx) || defined(STM32F7xx)
+      RX_DMA = { USART2, RCC_AHB1Periph_DMA1, 4, DMA1_Stream5 };
+    #endif
+    #if defined(STM32F0xx) || defined(STM32F1xx)
+      RX_DMA = { USART2, RCC_AHBPeriph_DMA1, DMA1, DMA1_Channel6 };
+    #endif
+  }
+  else if (peripheral == USART3) {
+    setRx(PIN_SERIAL3_RX);
+    setTx(PIN_SERIAL3_TX);
+    _uart_index = 2;
+
+    #if defined(STM32F2xx) || defined(STM32F4xx) || defined(STM32F7xx)
+      RX_DMA = { USART3, RCC_AHB1Periph_DMA1, 4, DMA1_Stream1 };
+    #endif
+    #ifdef STM32F1xx // F0 has no support for UART3, requires system remapping
+      RX_DMA = { USART3, RCC_AHBPeriph_DMA1, DMA1, DMA1_Channel3 };
+    #endif
+  }
+
+  #ifdef USART4 // must be F2 / F4 /F7
+    else if (peripheral == USART4) {
+      RX_DMA = { USART4, RCC_AHB1Periph_DMA1, 4, DMA1_Stream2 };
+      setRx(PIN_SERIAL4_RX);
+      setTx(PIN_SERIAL4_TX);
+      _uart_index = 3;
+    }
+  #endif
+
+  #ifdef UART4
+    else if (peripheral == UART4) {
+      #if defined(STM32F2xx) || defined(STM32F4xx) || defined(STM32F7xx)
+        RX_DMA = { UART4, RCC_AHB1Periph_DMA1, 4, DMA1_Stream2 };
+      #endif
+      #ifdef STM32F1xx // STM32F0xx has only 3 UARTs
+        RX_DMA = { UART4, RCC_AHBPeriph_DMA2, DMA2, DMA2_Channel3 };
+      #endif
+      setRx(PIN_SERIAL4_RX);
+      setTx(PIN_SERIAL4_TX);
+      _uart_index = 3;
+    }
+  #endif
+
+  #ifdef UART5 // must be F2 / F4 /F7
+    else if (peripheral == UART5) {
+      RX_DMA = { UART5, RCC_AHB1Periph_DMA1, 4, DMA1_Stream0 };
```

Comment:
This fails on F1 device.

Can it be updated with something like
`RX_DMA = { UART4, RCC_AHBPeriph_DMA2, DMA2, DMA2_Channel5 };`

or `DMA2_Channel4`? I'm not sure what the current usage of these is in Marlin

Reply:
> This fails on F1 device.

Thanks for the feedback. Which motherboard are you using?
What is the exact error message you get?

> Can it be updated with something like `RX_DMA = { UART4, RCC_AHBPeriph_DMA2, DMA2, DMA2_Channel5 };`

The UART4 code is present a few lines above already. A F1 device should not have a UART5, so the code should not cause issues. I could exclude the code for F1 devices if needed to solve this issue.

> or `DMA2_Channel4`? I'm not sure what the current usage of these is in Marlin

The DMA stream is linked to the UART port used, so that must be correct, it cannot freely be changed.

Reply:
I'm using `BOARD_MKS_ROBIN_E3_V1_1` and the issue is `DMA1_Stream0` does not exist for f1.

```
error: 'DMA1_Stream0' was not declared in this scope
```